test-engineering/contract/client/tests_openapi/test_openapi.py

```diff
@@ -0,0 +1,31 @@
+import os
+
+import schemathesis
+
+openapi_file = os.environ["OPENAPI_FILE"]
+contile_url = os.environ["CONTILE_URL"]
+schema = schemathesis.from_path(openapi_file, base_url=contile_url)
+
+
+def check_non_firefox_ua(response, case) -> None:
+    if response.status_code != 403:
+        raise AssertionError("Service did not return 403 for non-Firefox user-agent")
+
+
+def check_firefox_ua(response, case) -> None:
+    if response.status_code != 200:
+        raise AssertionError("Service did not return 200 for Firefox user-agent")
+
+
+@schema.parametrize()
+def test_api_from_firefox(case):
+    response = case.call_and_validate(
+        headers={"User-Agent": "Mozilla/5.0 (Windows NT 10.0; rv:10.0) Firefox/91.0"}
+    )
+    case.validate_response(response, additional_checks=(check_firefox_ua,))
+
+
+@schema.parametrize()
+def test_api_not_from_firefox(case):
+    response = case.call_and_validate()
+    case.validate_response(response, additional_checks=(check_non_firefox_ua,))
```

Comment:
Cool feature (additional_checks I mean)! Didn't know about this.

Note however in this case `test_api_not_from_firefox` will make hundreds of calls to `/v1/tiles` (as it fuzzes the input) only to check that we get a 403 in each case. That might be overkill?

Reply:
Running the pytests locally with the parameter ` --hypothesis-show-statistics` shows that only 1 test example is generated. Hypothesis can tell that the the search space has been exhausted.
```
=================================================================================== Hypothesis Statistics ====================================================================================
client/tests_openapi/test_openapi.py::test_api_from_firefox[GET /v1/tiles]:

  - during generate phase (0.09 seconds):
    - Typical runtimes: ~ 91ms, of which < 1ms in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - Stopped because nothing left to do


client/tests_openapi/test_openapi.py::test_api_not_from_firefox[GET /v1/tiles]:

  - during generate phase (0.00 seconds):
    - Typical runtimes: ~ 4ms, of which < 1ms in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - Stopped because nothing left to do
```